Enphase Envoy: add currents

pdeliot opened the pull request:
Added support phase currents, and three-phase with Enphase envoy #16744

pdeliot committed:
support phase currents, and three-phase with enphase envoy

pdeliot commented:
Yes, this was discussed with [ivoks](https://github.com/ivoks) on the issue discussion.
Originally I was returning 0 when phase is not available.
ivok suggested returning empty end tested it on its one phase system successfully.

This is the choosen way to implement 3 or 1 phase support.

andig commented:
We cannot simply return „nothing“.

andig commented:
We just cannot have this interface for 1p meters

pdeliot committed:
Update enphase.yaml replace "empty" value when missing by 0

pdeliot commented:
I've pushed version return 0 when value is not available instead on "empty".

andig commented:
@premultiply was meinst du? Gute Idee?

pdeliot commented:
What do you prefer?
-Return "empty"
-Return "0"
-Another way to deal with 1p and 3p

andig commented:
Imho we should not have 3p measurements on 1p devices at all, so 3rd option. Would any of the measurements (I guess current?) be missing if this was a 1p device?

pdeliot commented:
With 1p, as far as I know "powers", "currents" and "voltages" will not be supported.
But I have no docs about evcc requirement (if you have a pointer on the specifications, feel free to provide :-) )

I'm pushing a new version, with an optional parameter "phases".
Default is "1" then no behavior change for existing installations.

And you can set "3" to gain 3 phases support and linked capabilities.

pdeliot committed:
Adding a new way to support 1p and 3p

Adding "phases" optional parameter to switch on 3 phases support.

andig commented on templates/definition/meter/enphase.yaml:
```diff
@@ -30,6 +33,7 @@ render: |
     insecure: true 
     {{- end }}
     jq: .consumption[] | select(.measurementType == "net-consumption").wNow
+    {{- if eq .phases "3"}}
```
Please indent at the level of powers

pdeliot committed:
New way to support 3 phase

Added "phases" parameter to activate 3 phase support and attached capacities.

ivoks commented:
I'm worried that this would silently break existing 3p setups.

IMHO templates should collect as much info as possible and core engine should be able to cope with the collected data. If the template returns 'value none none' it should be obvious that it's a single phase system. If it returns 'value 0 0', that's a 3p system. I believe that using 0 in this patch is a wrong approach as you just can't know if the 0 is because there is no current or because there's no phase. Who knows what the future brings; evcc might eventually want to know if it's a 3p or 1p setup.

ivoks commented:
> [ivoks](https://github.com/ivoks) suggested returning empty and tested it on its one phase system successfully.

I've tested it on a 3p system without CTs ;)

andig commented:
To decide @premultiply:

- remove powers (only needed if currents not signed)
- remove voltages (only needed for chargers)
- allow phase measurements for 1p devices (and remove config option)

See https://github.com/evcc-io/docs/issues/650

pdeliot committed:
Update enphase.yaml

-removed powers
-removed voltages
-re-added 3 phase measurements for 1 phase devices
-removed 1/3 phase option

ivoks commented:
@pdeliot to work with both single and three phase system, you should really check the length of the array. For example, for phase in production:

curl http://${envoy_ip}/production.json?details=1 | jq 'if (( .production[] | select(.measurementType == "production").activeCount >= 1 ) and ( .production[] | select(.measurementType == "production").lines | length >= 1 )) then .production[] | select(.measurementType == "production").lines[2].rmsCurrent else 0 end'

but then for the second and third line:

curl http://${envoy_ip}/production.json?details=1 | jq 'if (( .production[] | select(.measurementType == "production").activeCount >= 1 ) and ( .production[] | select(.measurementType == "production").lines | length >= 3 )) then .production[] | select(.measurementType == "production").lines[2].rmsCurrent else 0 end'

And then similar for consumption. What do you think? I think this would cover all requirements; it would work only on CT enabled systems and it would work with both 1 and 3 phase systems without any parameters.

andig commented:
Ping @premultiply

premultiply commented:
@pdeliot Have you seen this post https://github.com/evcc-io/evcc/pull/17193#issuecomment-2511005056 ?

pdeliot commented:
I doubt this could be achieved...

It is allowing 1 or 3 phase detection, but at run time.
Then it would not allow "conditional" yaml creation.

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -27,9 +27,37 @@ render: |
+        password: {{ .token }}
+      insecure: true
+      {{- end }}
+      jq: if .consumption[] | select(.measurementType == "net-consumption").activeCount >= 1 then .consumption[] | select(.measurementType == "net-consumption").lines[0].rmsCurrent else 0 end
```
@pdeliot something like this:

jq: if (( .consumption[] | select(.measurementType == "net-consumption").activeCount >= 1 ) and ( .consumption[] | select(.measurementType == "net-consumption").lines | length >= 1 )) then .production[] | select(.measurementType == "net-consumption").lines[0].rmsCurrent else 0 end

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -27,9 +27,37 @@ render: |
+        password: {{ .token }}
+      insecure: true
+      {{- end }}
+      jq: if .consumption[] | select(.measurementType == "net-consumption").activeCount >= 1 then .consumption[] | select(.measurementType == "net-consumption").lines[1].rmsCurrent else 0 end
```
jq: if (( .consumption[] | select(.measurementType == "net-consumption").activeCount >= 1 ) and ( .consumption[] | select(.measurementType == "net-consumption").lines | length >= 1 )) then .consumption[] | select(.measurementType == "net-consumption").lines[1].rmsCurrent else 0 end

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -27,9 +27,37 @@ render: |
+        password: {{ .token }}
+      insecure: true
+      {{- end }}
+      jq: if .consumption[] | select(.measurementType == "net-consumption").activeCount >= 1 then .consumption[] | select(.measurementType == "net-consumption").lines[2].rmsCurrent else 0 end
```
jq: if (( .consumption[] | select(.measurementType == "net-consumption").activeCount >= 1 ) and ( .consumption[] | select(.measurementType == "net-consumption").lines | length >= 1 )) then .consumption[] | select(.measurementType == "net-consumption").lines[2].rmsCurrent else 0 end

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -53,6 +81,34 @@ render: |
+        password: {{ .token }}
+      insecure: true
+      {{- end }}
+      jq: if .production[] | select(.measurementType == "production").activeCount >= 1 then .production[] | select(.measurementType == "production").lines[0].rmsCurrent else 0 end
```
jq: if (( .production[] | select(.measurementType == "production").activeCount >= 1 ) and ( .production[] | select(.measurementType == "production").lines | length >= 1 )) then .production[] | select(.measurementType == "production").lines[0].rmsCurrent else 0 end

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -53,6 +81,34 @@ render: |
+        password: {{ .token }}
+      insecure: true
+      {{- end }}
+      jq: if .production[] | select(.measurementType == "production").activeCount >= 1 then .production[] | select(.measurementType == "production").lines[1].rmsCurrent else 0 end
```
jq: if (( .production[] | select(.measurementType == "production").activeCount >= 1 ) and ( .production[] | select(.measurementType == "production").lines | length >= 1 )) then .production[] | select(.measurementType == "production").lines[1].rmsCurrent else 0 end

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -53,6 +81,34 @@ render: |
+        password: {{ .token }}
+      insecure: true
+      {{- end }}
+      jq: if .production[] | select(.measurementType == "production").activeCount >= 1 then .production[] | select(.measurementType == "production").lines[2].rmsCurrent else 0 end
```
jq: if (( .production[] | select(.measurementType == "production").activeCount >= 1 ) and ( .production[] | select(.measurementType == "production").lines | length >= 1 )) then .production[] | select(.measurementType == "production").lines[2].rmsCurrent else 0 end

ivoks commented on templates/definition/meter/enphase.yaml:
```diff
@@ -27,9 +27,37 @@ render: |
       type: bearer
       password: {{ .token }}
-    insecure: true
+    insecure: true 
```
Trailing space, should be removed.

premultiply replied:
```suggestion
    insecure: true
```

ivoks reviewed (changes requested):
This is what I had in mind. But please also be aware that there are Enphase firmwares out there that do not list meters even in detailed view (see bug 17817). You will have to merge your changes with the fix that just landed...

pdeliot committed:
Merge branch 'master' into patch-1

premultiply committed:
remove ws
